affinity/cpp-20/d0796r1.md

```diff
@@ -136,25 +142,54 @@ There are some additional challenges which we have been investigating but are no
 
     /* Execution context */
 
-    struct execution_context {
+    class execution_context {
+     public:
 
       using executor_type = __unspecfied__;
 
-      template <typename ExecutionResource>
-      execution_context(ExecutionResource &&execResource);
+      execution_context(const execution_resource &);
 
       ~execution_context();
 
       const execution_resource &resource() const noexcept;
 
       executor_type executor() noexcept;
 
+      pmr::memory_resource *memory_resource() noexcept;
+
+    };
+
+    /* Affinity query */
+
+    enum class affinity_operation { read, write, copy, move, map };
+    enum class affinity_metric { latency, bandwidth, capacity, power_consumption };
+
+    template <affinity_operation Operation, affinity_metric Metric>
+    class affinity_query {
+     public:
+
+      using native_affinity_type = __unspecified__;
+      using error_type = __unspecified__
+
+      affinity_query(execution_resource &&, execution_resource &&);
+
+      ~affinity_query();
+
+      native_affinity_type native_affinity() const noexcept;
+
+      friend expected<size_t, error_type> operator==(const affinity_query&, const affinity_query&);
+      friend expected<size_t, error_type> operator!=const affinity_query&, const affinity_query&);
+      friend expected<size_t, error_type> operator<(const affinity_query&, const affinity_query&);
+      friend expected<size_t, error_type> operator>(const affinity_query&, const affinity_query&);
+      friend expected<size_t, error_type> operator<=(const affinity_query&, const affinity_query&);
+      friend expected<size_t, error_type> operator>=(const affinity_query&, const affinity_query&);
+
     };
 
     /* This system */
 
     namespace this_system {
-      const execution_resource &resource();
+      std::vector<execution_resource> resources() noexcept;
```

Comment:
My expectation with the current approach would be that if a dynamic library introduces a new device at runtime it would not affect the topology as seen by `this_system::resources()`.

Reply:
I think a dynamic library should be able to populate the system resources before the first call (notionally before main, but I guess is easier to define before the first call to resources). Otherwise how the dynamic loading of OpenCL would load the icd, for example?

Reply:
So this is what the paper currently says, sorry my last comment wasn't very clear. By introducing a new device at runtime I meant after `main`, so with the current wording, any runtime is permitted to modify the system topology as seen by `this_system::resources()` as long as it happens before `main` is called. We may want to loosen that requirement to the first call to `this_system::resources()` in the future if this is problematic, but for now I think this is fine.